onnxruntime/core/providers/rocm/nn/pool.cc

```diff
@@ -0,0 +1,353 @@
+    ORT_RETURN_IF_ERROR(y_tensor.Set(y_dims_miopen, MiopenTensor::GetDataType<HipT>()));
+
+    MIOPEN_RETURN_IF_ERROR(PoolingForwardHelper(GetMiopenHandle(context), pooling_desc, &alpha, x_tensor, x_data, &beta, y_tensor, y_data));
+  }
+
+  return Status::OK();
+}
+
+template <typename T>
+Status Pool<T, MaxPool<8>>::ComputeInternal(OpKernelContext* context) const {
+  typedef typename ToHipType<T>::MappedType HipT;
+  const Tensor* X = context->Input<Tensor>(0);
+  const TensorShape& x_shape = X->Shape();
+
+  if (x_shape.NumDimensions() < 3) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Input dimension cannot be less than 3.");
+  }
+
+  auto kernel_shape = this->pool_attrs_.kernel_shape;
+  auto pads = this->pool_attrs_.pads;
+  auto strides = this->pool_attrs_.strides;
+
+  if (this->pool_attrs_.global_pooling) {
+    ORT_ENFORCE("unreachable");
+  }
+
+  auto y_dims = this->pool_attrs_.SetOutputSize(x_shape, x_shape[1], &pads);
+  Tensor* Y = context->Output(0, TensorShape(y_dims));
+
+  // special case when there is a dim value of 0 in the shape.
+  if (Y->Shape().Size() == 0)
+    return Status::OK();
+
+  auto x_data = reinterpret_cast<const HipT*>(X->Data<T>());
+  auto y_data = reinterpret_cast<HipT*>(Y->MutableData<T>());
+
+  Tensor* I = context->Output(1, TensorShape(y_dims));
+  if (nullptr != I || !this->pool_attrs_.default_dilations) {
+    auto i_data = nullptr == I ? nullptr : I->MutableData<int64_t>();
+    MaxPoolWithIndex<HipT>(
+        this->Stream(context),
+        x_shape,
+        TensorShape(y_dims),
+        kernel_shape,
+        strides,
+        pads,
+        this->pool_attrs_.dilations,
+        this->pool_attrs_.storage_order,
+        x_data,
+        y_data,
+        i_data);
+  } else {
+    ORT_RETURN_IF_ERROR((Pool<T, MaxPool<1>>::ComputeInternal(context)));
+  }
+  return Status::OK();
+}
+
+template <typename T, typename PoolType>
+Status GlobalPool<T, PoolType>::ComputeInternal(OpKernelContext* context) const {
+  using HipT = typename ToHipType<T>::MappedType;
+  const Tensor* X = context->Input<Tensor>(0);
+  const TensorShape& x_shape = X->Shape();
+
+  if (x_shape.NumDimensions() < 3) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Input dimension cannot be less than 3.");
+  }
+
+  ORT_ENFORCE(this->pool_attrs_.global_pooling);
+
+  miopenReduceTensorOp_t reduce_op;
+  if constexpr (PoolType::type == onnxruntime::PoolType::kAveragePool) {
+    reduce_op = MIOPEN_REDUCE_TENSOR_AVG;
+  } else if (PoolType::type == onnxruntime::PoolType::kMaxPool) {
+    reduce_op = MIOPEN_REDUCE_TENSOR_MAX;
+  } else {
+    ORT_NOT_IMPLEMENTED();
+  }
+
+  miopenDataType_t miopen_type_X = MiopenTensor::GetDataType<HipT>();
+
+  MiopenReduceDescriptor reduce_desc;
+  if constexpr (std::is_same<T, MLFloat16>::value) {
+    ORT_RETURN_IF_ERROR(reduce_desc.Set(
+        reduce_op, MiopenTensor::GetDataType<float>(), MIOPEN_REDUCE_TENSOR_FLATTENED_INDICES));
```

Comment:
Why use `MiopenTensor::GetDataType<float>()` here?

Reply:
Copied from Reduce* Op impl. It should just use acc type as fp32 for numerical stability.